libs/ml/examples/word2vec_distributed_learning/main.cpp

```diff
@@ -0,0 +1,185 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/random.hpp"
+#include "file_loader.hpp"
+#include "math/clustering/knn.hpp"
+#include "math/matrix_operations.hpp"
+#include "math/statistics/mean.hpp"
+#include "math/tensor.hpp"
+#include "ml/core/graph.hpp"
+#include "ml/dataloaders/word2vec_loaders/sgns_w2v_dataloader.hpp"
+#include "ml/distributed_learning/coordinator.hpp"
+#include "ml/layers/fully_connected.hpp"
+#include "ml/layers/skip_gram.hpp"
+#include "ml/ops/activation.hpp"
+#include "ml/ops/loss_functions.hpp"
+#include "ml/optimisation/sgd_optimiser.hpp"
+#include "model_saver.hpp"
+#include "word2vec_client.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <ctime>
+#include <fstream>
+#include <iostream>
+#include <list>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#define NUMBER_OF_CLIENTS 3
+#define NUMBER_OF_ITERATIONS 100
+#define NUMBER_OF_ROUNDS 10
+#define SYNCHRONIZATION_MODE CoordinatorMode::ASYNCHRONOUS
+
+#define BATCH_SIZE 128
+#define LEARNING_RATE .001f
+#define TEST_SET_RATIO 0.03f
+#define NUMBER_OF_PEERS 2
+
+using namespace fetch::ml::ops;
+using namespace fetch::ml::layers;
+using namespace fetch::ml;
+using namespace fetch::ml::dataloaders;
+
+using DataType         = double;
+using TensorType       = fetch::math::Tensor<DataType>;
+using VectorTensorType = std::vector<TensorType>;
+using SizeType         = typename TensorType::SizeType;
+
+// std::string ReadFile(std::string const &path)
+//{
+//  std::ifstream t(path);
+//  return std::string((std::istreambuf_iterator<char>(t)), std::istreambuf_iterator<char>());
+//}
+
+int main(int ac, char **av)
+{
+  if (ac < 2)
+  {
+    std::cout << "Usage : " << av[0] << " PATH/TO/text8" << std::endl;
+    return 1;
+  }
+
+  std::string train_file = av[1];
+
+  std::shared_ptr<Coordinator> coordinator =
+      std::make_shared<Coordinator>(SYNCHRONIZATION_MODE, NUMBER_OF_ITERATIONS);
+
+  std::cout << "FETCH Distributed Word2vec Demo -- Asynchronous" << std::endl;
+  std::srand((unsigned int)std::time(nullptr));
```

Comment:
don't use unsigned int - specify the size of the object: e.g. SizeType, uint32_t, etc.

Reply:
This line shouldn't even exist, because we are using fetch:: random.